cgfs.py

```diff
@@ -102,36 +106,147 @@ def pop(self, filename):
         return file
 
     def read(self):
-        return ['.', '..'] + [x for x in self.children]
+        res = list(self.children)
+        res.extend(['.', '..'])
+        return res
+
+
+class TempDirectory(Directory):
+    def __init__(self, *args, **kwargs):
+        super(TempDirectory, self).__init__(*args, **kwargs)
+        self.stat = {
+            'st_size': 0,
+            'st_atime': time(),
+            'st_ctime': time(),
+            'st_size': 0,
+            'st_mtime': time(),
+            'st_mode': S_IFDIR | 0o770,
+            'st_nlink': 2,
+        }
+
+
+class TempFile:
+    def __init__(self, name, tmpdir):
+        self._tmpdir = tmpdir
+        self.name = name
+
+        self.__handle = None
+
+        # Create a new temporary file
+        self._filename = str(uuid.uuid4())
+
+        while path.exists(
+            path.join(path.join(self._tmpdir, self._filename)),
+        ):  # pragma: no cover
+            self._filename = str(uuid.uuid4())
+
+        self.full_path = path.join(self._tmpdir, self._filename)
+        Path(path.join(self.full_path)).touch()
+
+    @property
+    def stat(self):
+        st = os.lstat(self.full_path)
+        stat = {
+            key: getattr(st, key)
+            for key in (
+                'st_atime', 'st_ctime', 'st_gid', 'st_mode', 'st_mtime',
+                'st_nlink', 'st_size', 'st_uid'
+            )
+        }
+        stat['st_mode'] = S_IFREG | 0o770
+        return stat
+
+    @property
+    def _handle(self):
+        if self.__handle is None:
+            self.__handle = open(self.full_path, 'r+b')
+            self.__handle.seek(0)
+        return self.__handle
+
+    def getattr(self):
+        return self.stat
+
+    def setattr(self, key, value):  # pragma: no cover
+        raise ValueError
+
+    def utimens(self, atime, mtime):
+        os.utime(self.full_path, (atime, mtime))
+
+    def open(self, *args):
+        # We open on-demand so we never have problems with opening
+        pass
+
+    def read(self, offset, size):
+        self._handle.seek(offset)
+        return self._handle.read(size)
+
+    def write(self, data, offset):
+        handle = self._handle
+        handle.seek(offset)
+        res = handle.write(data)
+        handle.flush()
+        return res
+
+    def release(self):
+        if self.__handle is not None:
+            self.__handle.flush()
+            self.__handle.close()
+        self.__handle = None
```

Comment:
We should keep a counter how many times a file is `open`ed and only close the handle when the counter becomes 0. A file can still be open in a process after it has been `unlink`ed, but if we close the file on each `release` and `unlink` was called on the file we have no way to recover the file contents anymore.